### ISSUE TITLE
Mac OS X update breaks pymediainfo

### ISSUE THREAD

**Issue by mattbodman** [sbraz/pymediainfo#24]
Could have been the update to OS X 10.11.4 or an update of Command Line Tools (sorry I'm not sure which caused it).  However, using CDLL no longer works.. This error results:  OSError: dlopen libmediainfo.so.0, 6 cannot be found, so I guess Apple moved something.  This exact happened after the upgrades on two separate Apple machines and the fix below worked for both.

To fix, I rolled back pymediainfo to v 1.4

**Comment by sbraz** [comment]
Are you using the git version or the 2.0 release?

**Comment by sbraz** [comment]
If you use the git version, it should never attempt to load the `.so`: https://github.com/sbraz/pymediainfo/commit/cb0f6f40abcbb4f3ffe518c794668206b101b0af#diff-78f1fb93e3c37202e83112bf2277293dR70

**Comment by mattbodman** [comment]
Ok thanks.  Yes, it was v2.0 that pip installed.

**Comment by sbraz** [comment]
Can you please try the git version? `pip install git+https://github.com/sbraz/pymediainfo.git`
I have not made a new release yet as there are still quite a few bugs around.

**Comment by mattbodman** [comment]
Still didn't work.  Here is the error traceback:

`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Volumes/Extra/Repos/env/lib/python2.7/site-packages/pymediainfo/__init__.py", line 72, in parse
    lib = CDLL("libmediainfo.0.dylib")
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ctypes/__init__.py", line 365, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(libmediainfo.0.dylib, 6): image not found`

**Comment by sbraz** [comment]
Is it called something else on your system?

**Comment by mattbodman** [comment]
There's this file:

`/Applications/MediaInfo.app/Contents/Resources/libmediainfo.dylib`

**Comment by sbraz** [comment]
That's weird, it means the official Python bindings have it wrong…
https://github.com/MediaArea/MediaInfoLib/blob/master/Source/MediaInfoDLL/MediaInfoDLL.py#L34
I don't have a Mac computer but I should be able to get my hands on one at some point. I really need to do it as it seems many people use the package on Mac OS.

**Comment by sbraz** [comment]
There is a `libmediainfo.0.dylib` file in http://mediaarea.net/download/binary/libmediainfo0/0.7.84/MediaInfo_DLL_0.7.84_Mac_i386+x86_64.tar.bz2, please make sure this is the version you are using.

**Comment by mattbodman** [comment]
Ah ok.  I was using, and have always just installed the CLI version because I use MediaInfo outside Python as well.  Pymediainfo traditionally worked with the CLI, obviously.  Maybe its worth mentioning the DyLib version in your install docs or something? (if you haven't already)

**Comment by sbraz** [comment]
Since I took over development (version 2.0), the command-line application is no longer used. The [changelog](https://github.com/sbraz/pymediainfo/releases/tag/v2.0) mentions it but maybe not clearly enough.
You can still parse XML generated by the CLI version if you want to:
```python
pymediainfo.MediaInfo(subprocess.check_output(["mediainfo", "--output=XML", "-f", "filename.ext"]))
```
Using the library is just easier and more efficient.
I will try to make the documentation more clear and I'll release a new version as soon as you can confirm that the git version now works on your setup.

**Comment by mattbodman** [comment]
Ok, makes sense.  I'm actually unsure where to put the DyLib file.  Do you know?

**Comment by sbraz** [comment]
I have little experience with Mac OS, did you manage to make it work? It probably goes to /lib or similar.

**Comment by sbraz** [comment]
Hi @mattbodman, did you manage to load the dylib properly? Can we close this issue?

**Comment by mattbodman** [comment]
Sorry, haven't had a chance to try it. Close it if you like

**Comment by mattbodman** [comment]
I actually ran out of time to figure out where to put the dylib file, so I reverted to Linux.  Sorry about that.

**Comment by sbraz** [comment]
Hi!
Where is the dylib located?

**Comment by DanielsWrath** [comment]
Hey! The file ```libmediainfo.dylib``` is in ```/Applications/MediaInfo.app/Contents/Resources```. I downloaded it from the app store. I believe this is the one you wanted to know the location of?

**Comment by sbraz** [comment]
I'm no Mac expert but I think that simply overriding `LD_LIBRARY_PATH` to include the containing folder would fix the issue. Can you try adding this before you load pymediainfo?
```py
os.environ["LD_LIBRARY_PATH"] = os.environ.get("LD_LIBRARY_PATH", "") +  ":/Applications/MediaInfo.app/Contents/Resource"
```

**Comment by DanielsWrath** [comment]
Nope, sorry still doesn't work, the error is the same. I have solved my program without using pymediainfo though.

**Comment by sbraz** [comment]
What about
```py
os.environ["DYLD_LIBRARY_PATH"] = os.environ.get("DYLD_LIBRARY_PATH", "") +  ":/Applications/MediaInfo.app/Contents/Resource"
```

**Comment by DanielsWrath** [comment]
Nope sorry, still the same error

**Comment by OmgImAlexis** [comment]
Are there any updates with this?

**Comment by sbraz** [comment]
@OmgImAlexis I don't really know, I can't reproduce it myself as I don't have a Mac to test all the time.
Does it work if you point `DYLD_LIBRARY_PATH` to the right location and export it before running a Python programme?

**Comment by OmgImAlexis** [comment]
Yes and also works if we use `DYLD_FALLBACK_LIBRARY_PATH`.

We also tried editing pymediainfo and adding this before the command to try CDLL. From the looks of it the ENV param needs to be set before Python starts, I wonder if you could get around that by starting it and then forking the process or does Python itself need restarting?

```
os.environ['DYLD_FALLBACK_LIBRARY_PATH'] = os_folder + ':' + os.environ.get('DYLD_FALLBACK_LIBRARY_PATH', '')
```

os_folder is the absolute path to our copy of `libmediainfo.0.dylib`.

**Comment by OmgImAlexis** [comment]
@ratoaq2

**Comment by sbraz** [comment]
Nice, thanks for testing! I'll borrow someone's Mac and perform more tests. Maybe ctypes has a parameter to add more paths from which to load its libraries.

**Comment by sbraz** [comment]
What if you set the environment variable before importing ctypes, does it help?

**Comment by OmgImAlexis** [comment]
That does seem to work. The problem would be getting the correct path though since it could be anywhere the user has added it especially if the program using it ships with mediainfo and the dylib file.

**Comment by OmgImAlexis** [comment]
Maybe allow a method to set the path if people need to override it?

**Comment by sbraz** [comment]
Well the programme can look for the dylib in several paths and pass them to
pymediainfo. I'll probably add an optional parameter to choose what path to
load the dylib from. Your code could then use it, what do you think?

**Comment by OmgImAlexis** [comment]
Yeah, that should be fine. @ratoaq2 can you confirm?

**Comment by sbraz** [comment]
@OmgImAlexis Can you try replacing `lib = CDLL("libmediainfo.0.dylib")` with `cdll.LoadLibrary(absolute

**Comment by bviews** [comment]
I have the same issue when I use GUI version from Mac App Store.

then I download the CLI version form https://mediaarea.net/nn/MediaInfo/Download/Mac_OS and problem solved

**Comment by sbraz** [comment]
@bviews where does the app install the .dylib?

**Comment by GiancarloFusiello** [comment]
Download and extract the files @sbraz mentioned previously to `~/Downloads`: http://mediaarea.net/download/binary/libmediainfo0/0.7.84/MediaInfo_DLL_0.7.84_Mac_i386+x86_64.tar.bz2

Then copy the file `libmediainfo.0.dylib` to `/usr/local/lib`

**Comment by djiesamsoe** [comment]
@GiancarloFusiello thanks, that work for me.

**Comment by sound2picture** [comment]
libmediainfo.0.dylib to /usr/local/lib solution works! Thank you!

**Comment by rebelaide** [comment]
libmediainfo.0.dylib to /usr/local/lib solution worked for me too.

**Comment by sensharma** [comment]
@GiancarloFusiello You're a start. I had the same issue on Mac. Chanced upon this thread on searching and that solved it for me. For my understanding, is the problem with the MediaInfo app store app (that's where I installed it from)?

**Comment by fralomba** [comment]
It doesn't work for me...any other idea? the error continues being: "OSError: dlopen(libmediainfo.dylib, 6): image not found"

**Comment by sbraz** [comment]
@fralomba where is your lib located?

**Comment by fralomba** [comment]
I have installed it with pip, and tried to follow @GiancarloFusiello 's hint but nothing changed. Same error. The lib is located in usr/local/lib. Do u know how to help me?

**Comment by patelsp** [comment]
@fralomba Here is what I did & works ...

Download & Extract following
1) https://mediaarea.net/en/MediaInfo/Download/Mac_OS/v0.7.99 (CLI version)
2) http://mediaarea.net/download/binary/libmediainfo0/0.7.84/MediaInfo_DLL_0.7.84_Mac_i386+x86_64.tar.bz2

Create directory
3) sudo mkdir /usr/local/lib   NOTE: Can't use "/usr/lib" is protected by SIP
4) copy extracted libmediainfo.0.dylib into /usr/local/lib 
    copy same file with name as libmediainfo.dylib

Modify
5) /Users/<username>/Downloads/pymediainfo-2.1.9/pymediainfo/__init__.py
line 79 change to return CDLL("/usr/local/lib/libmediainfo.0.dylib")
line 81 change to return CDLL("/usr/local/lib/libmediainfo.dylib")

Install
6) cd /Users/<username>/Downloads/pymediainfo-2.1.9/
- python setup.py build
- python setup.py install

**Comment by sbraz** [comment]
@fralomba Just to be sure: you're aware that pip does not install the mediainfo library, right?

**Comment by sbraz** [comment]
There should be no need to move dylib files around. With the latest version, you can simply pass the dylib path to `parse`'s `library_file` parameter:
https://github.com/sbraz/pymediainfo/commit/05ae1896a132edd8e2aa61c963c7ab36c9f356ca#diff-78f1fb93e3c37202e83112bf2277293dR101

**Comment by joyearnaud** [comment]
Hi, 
I have OS Big Sur 11.2.2, I have tried out all the solutions in this thread but still getting the error

```
Traceback (most recent call last):
  File "/Users/xxx/Project/filesort/run.py", line 11, in <module>
    sort_files(PATH, length = LENGTH)
  File "/Users/xxx/Project/filesort/filesort/sort_files.py", line 36, in sort_files
    sort_other(filepath, out_repo)
  File "/Users/xxx/Project/filesort/filesort/sort_files.py", line 167, in sort_other
    if is_video(filepath, out_repo):
  File "/Users/xxx/Project/filesort/filesort/sort_files.py", line 157, in is_video
    media_info = MediaInfo.parse(filepath, library_file="/usr/local/lib/libmediainfo.dylib")
  File "/Users/xxx/Library/Python/3.8/lib/python/site-packages/pymediainfo/__init__.py", line 422, in parse
    lib, handle, lib_version_str, lib_version = cls._get_library(library_file)
  File "/Users/xxx/Library/Python/3.8/lib/python/site-packages/pymediainfo/__init__.py", line 324, in _get_library
    raise OSError("Failed to load library")
```

**Comment by sbraz** [comment]
@joyearnaud I have just pushed a new branch with more verbose exceptions when the library cannot be loaded, can you try it with
```
pip install git+https://github.com/sbraz/pymediainfo@library_detection
```
And let me know what the exception looks like now? It should show something like this with the path that were tested:
```
OSError: Failed to load library from /usr/local/lib/libmediainfo.dylib - /usr/local/lib/libmediainfo.dylib: <some error here>
```

**Comment by joyearnaud** [comment]
Thank for your prompt answer

#### My PATH
```BASH
% echo "$PATH"
/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/homebrew/bin
```

#### Medialib files location
```BASH
% ls usr/local/lib 
libmediainfo.0.dylib    libmediainfo.dylib
% ls usr/local/bin
libmediainfo.0.dylib    libmediainfo.dylib      mediainfo
```

#### Error looks like with new verbose branch
```BASH
Traceback (most recent call last):
  File "/Users/xxx/Project/filesort/run.py", line 11, in <module>
    sort_files(PATH, length = LENGTH)
  File "/Users/xxx/Project/filesort/filesort/sort_files.py", line 36, in sort_files
    sort_other(filepath, out_repo)
  File "/Users/xxx/Project/filesort/filesort/sort_files.py", line 167, in sort_other
    if is_video(filepath, out_repo):
  File "/Users/xxx/Project/filesort/filesort/sort_files.py", line 157, in is_video
    media_info = MediaInfo.parse(filepath)
  File "/Users/xxx/Library/Python/3.8/lib/python/site-packages/pymediainfo/__init__.py", line 431, in parse
    lib, handle, lib_version_str, lib_version = cls._get_library(library_file)
  File "/Users/xxx/Library/Python/3.8/lib/python/site-packages/pymediainfo/__init__.py", line 329, in _get_library
    raise OSError(
OSError: Failed to load library from libmediainfo.0.dylib, libmediainfo.dylib - dlopen(libmediainfo.0.dylib, 6): no suitable image found.  Did find:
        libmediainfo.0.dylib: no matching architecture in universal wrapper
        /usr/local/lib/libmediainfo.0.dylib: no matching architecture in universal wrapper
        /usr/local/lib/libmediainfo.0.dylib: no matching architecture in universal wrapper
        /Users/xxx/Project/filesort/libmediainfo.0.dylib: no matching architecture in universal wrapper, dlopen(libmediainfo.dylib, 6): no suitable image found.  Did find:
        libmediainfo.dylib: no matching architecture in universal wrapper
        /usr/local/lib/libmediainfo.dylib: no matching architecture in universal wrapper
        /usr/local/lib/libmediainfo.0.dylib: no matching architecture in universal wrapper
        /Users/xxx/Project/filesort/libmediainfo.0.dylib: no matching architecture in universal wrapper
```

**Comment by sbraz** [comment]
I'm no Mac OS expert but it looks like your library isn't compatible with the version you have, do the MediaInfo GUI or the CLI (`mediainfo`) work?

**Comment by sbraz** [comment]
@joyearnaud [It looks like you need version 21.03 of the library](https://github.com/MediaArea/MediaInfoLib/blob/d5fb067e1539aa7a74c491e8262c81214f9c8bcb/History_DLL.txt#L25) for Apple Silicon, is that the one you're using?

**Comment by joyearnaud** [comment]
> I'm no Mac OS expert but it looks like your library isn't compatible with the version you have, do the MediaInfo GUI or the CLI (`mediainfo`) work?

Seems to work
```BASH
xxx@MacBook bin % mediainfo /Users/xxx/Data  
General
Complete name                            : /Users/xxx/Data/P1090707.MOV
Format                                   : QuickTime
Format/Info                              : Original Apple specifications
File size                                : 11.4 MiB
```

> [It looks like you need version 21.03 of the library](https://github.com/MediaArea/MediaInfoLib/blob/d5fb067e1539aa7a74c491e8262c81214f9c8bcb/History_DLL.txt#L25) for Apple Silicon, is that the one you're using?
```BASH

xxx@MacBook bin % mediainfo --version            
MediaInfo Command line, 
MediaInfoLib - v21.03

```

**Comment by sbraz** [comment]
I don't think Python 3.8 is supposed to support Apple Silicon (or OSX1 11), where did you get it?
The downloads page for the latest 3.8 release doesn't mention OSX 11 or Apple silicon:
https://www.python.org/downloads/release/python-388/
But the page for 3.9 mentions a specific installer for it:
https://www.python.org/downloads/release/python-394/

Can you try with this newer Python installer please?

**Comment by joyearnaud** [comment]
@sbraz I'm brand new Apple user and it's a brand new MacBook: Python 2.xx and Python 3.8 are preinstalled;

Browsing about this version problem show me this [article](https://scriptingosx.com/2020/02/wrangling-pythons/):

> This is great, right? Apple says there is a built-in Python 3! And it’s pre-installed? Just move all your scripts to Python 3 and you’ll be fine!

> Unfortunately, not quite. The python3 binary does exist on a ‘clean’ macOS, but it is only a stub tool, that will prompt a user to download and install the Command Line Developer Tools (aka “Developer Command Line Tools” or “Command Line Tools for Xcode”). This is common for many tools that Apple considers to be of little interest to ‘normal,’ non-developer users. Another common example is git.

I will definitively try version 3.9 and let you know.

**Comment by joyearnaud** [comment]
@sbraz Thank for your time and advice, Python 3.9.4 works fine and resolves my problem

**Comment by sbraz** [comment]
You're welcome. Can you tell me if the pymediainfo wheel (which includes libmediainfo) works out-of-the box or if you had to put the `dylib` somewhere?

Since the lib is built with v20.09, I doubt it'll work. In that case, I might create a new release that includes the newer libmediainfo.

**Comment by joyearnaud** [comment]
> You're welcome. Can you tell me if the pymediainfo wheel (which includes libmediainfo) works out-of-the box or if you had to put the `dylib` somewhere?
> 
> Since the lib is built with v20.09, I doubt it'll work. In that case, I might create a new release that includes the newer libmediainfo.

First I put everything.

Now with this conf (without `dylib`)
```BASH
xxx@MacBook filesort % ls -R /usr/local  
bin	lib

/usr/local/bin:
2to3			easy_install-3.9	idle3.9			pip3			pydoc3			python3			python3.9
2to3-3.9		idle3			mediainfo		pip3.9			pydoc3.9		python3-config		python3.9-config

/usr/local/lib:
 
```
it's working too.

**Comment by sbraz** [comment]
Hmm so the older version did work with Apple Silicon then. Does `python -c "from pymediainfo import MediaInfo; print(MediaInfo._get_library())"` return something like this?
```
(<CDLL 'libmediainfo.dylib', handle 5623796306e0 at 0x7f16ae6e7e80>, 94710359423712, '20.09', (20, 9))
```

**Comment by joyearnaud** [comment]
**erratum** : In my process to make it work I put `dylib` in my project, that seems to be necessary;

With these `dylib` in my root working dir
```BASH
xxx@MacBook filesort % python3 -c "from pymediainfo import MediaInfo; print(MediaInfo._get_library())"
(<CDLL 'libmediainfo.0.dylib', handle 7fb743ca8ad0 at 0x7fb7481361c0>, 140425093475744, '0.7.84', (0, 7, 84))
```

So, deleting them made running your command fail with 
```BASH
xxx@MacBook filesort % python3 -c "from pymediainfo import MediaInfo; print(MediaInfo._get_library())"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/xxx/Library/Python/3.9/lib/python/site-packages/pymediainfo/__init__.py", line 329, in _get_library
    raise OSError(
OSError: Failed to load library from libmediainfo.0.dylib, libmediainfo.dylib - dlopen(libmediainfo.0.dylib, 6): image not found, dlopen(libmediainfo.dylib, 6): image not found
```

**Comment by sbraz** [comment]
Can we maybe chat on https://gitter.im/ to avoid flooding this issue? I'd like to make it easier for future pymediainfo users to run it on OSX, and from what I see:
* either you didn't install it from the OSX wheel
* or the dylib from the wheel doesn't work (but your old 0.7.x one does which would be weird)

**Comment by swannie-eire** [comment]
for anyone else having this issue. I encountered it on 12.5.1 on M1/arm mac.

The only thing that worked for me was downloading the libmediainfo.0.dylib, libmediainfo.dylib from here https://mediaarea.net/en/MediaInfo/Download/Mac_OS

**Comment by sbraz** [comment]
Hi @swannie-eire so the bundled MediaInfo v21.03 does not work but newer versions do? Can you please try older versions and let me know which version fixed the problem. I'll probably make a new pymediainfo release once I know that this actually fixes the problem.

**Comment by sbraz** [comment]
Hi @JeromeMartinez thanks for the feedback, I'll check that. This means I should create two binary packages? One for M1 Macs and one for older CPUs?

**Comment by JeromeMartinez** [comment]
> One for M1 Macs and one for older CPUs?

No. One for i386 and one for newer CPUs is a lot better, as i386 is maybe for 0.1% (or less!) of your users, it would be less problematic to have an unique build for 99.9% rather that a split between ARM and Intel CPU (still lot of x86_64 users). Or maybe just avoid an i386 version, last macOS with i386 support is 10.6 (11 year old), as we did (v21.09+ are without i386 support, we switched from i386+x86_64 to x86_64+arm64 because in 2021 nearly no one uses new versions of MediaInfo on 10+ year old machines).

but I am surprised that you had no complain about M1/arm compatibility sooner, such machine is very common nowadays, maybe users use pymediainfo with another version of MediaInfo library than yours?
Generally speaking, from MediaInfo upstream point of view we would prefer that you update the "default" (in your repo) version of MediaInfo library is more often updated for taking advantage of the new formats we support, but we also understand that QA would be bigger on your side so this is understandable that you keep using an older version, in that case switching  from [i386+x86_64](https://mediaarea.net/download/binary/libmediainfo0/21.03/MediaInfo_DLL_21.03_Mac_i386+x86_64.tar.bz2) to [x86_64+arm64](https://mediaarea.net/download/binary/libmediainfo0/21.03/MediaInfo_DLL_21.03_Mac_x86_64+arm64.tar.bz2) is the less risky.

**Comment by sbraz** [comment]
Thanks for the explanation @JeromeMartinez, I have to admit I kinda forgot that the new CPUs were using a different architecture. I have just released pymediainfo 6.0.0 and I built the wheel using the arm64-compatible libmediainfo → https://files.pythonhosted.org/packages/75/1a/2f018bfe7678fa48e1f4d30b2bba77913bd5145966b405b383e2716093e1/pymediainfo-6.0.0-py3-none-macosx_12_0_x86_64.whl
I also performed the build on a more modern macOS version so the file ends up being named `12_0` instead of `10_9`. I hope that won't cause issues to consumers.
